### PR TITLE
Generate dedicated Context for errors / Align capture_body config values

### PR DESCRIPTION
**Pull request by mikker**
Right now, errors get their `Context` copied over from `current_transaction` which isn't necessarily the correct dat, and config options might differ for the two different event types.

This PR makes sure we generate separate, new contexts for errors only copying over relevant data such as `tags` and `custom`.

It started out as aligning on the values for the option `capture_body` so that's in there too.

### CODE REVIEW

File: lib/elastic_apm.rb [hunk 1]
```diff
@@ -285,27 +285,43 @@ def with_span(
...
     # @param rack_env [Rack::Env] A Rack env
     # @return [Context] The built context
-    def build_context(rack_env)
-      agent&.build_context(rack_env)
+    def build_context(
+      deprecated_env = nil,
+      rack_env: nil,
+      for_type: :transaction
+    )
+      if !rack_env && (rack_env = deprecated_env)
```
Comment (simitt) [inline lib/elastic_apm.rb]: Why did you introduce `deprecated_env` here and make `rack_env` a keyword arg but then require it?
How about `def build_context(rack_env:, for_type:)` instead?

File: lib/elastic_apm/config.rb [hunk 2]
```diff
@@ -296,6 +296,31 @@ def collect_metrics?
...
+
+      case value
+      when true
+        alert_logger.warn "Boolean value for option `capture_body' has " \
+          "been deprecated. Setting to 'all'"
+        @capture_body = 'all'
+      when false
+        alert_logger.warn "Boolean value for option `capture_body' has " \
+          "been deprecated. Setting to 'off'"
+        @capture_body = 'off'
```
Comment (simitt) [inline lib/elastic_apm/config.rb]: 👍

File: lib/elastic_apm/context.rb [hunk 3]
```diff
@@ -9,13 +9,24 @@
...
+    attr_accessor :user
+    attr_reader :custom
+    attr_reader :tags
+
+    def empty?
+      return false if tags.any?
+      return false if custom.any?
+      return false unless user.empty?
+
+      true
```
Comment (simitt) [inline lib/elastic_apm/context.rb]: You can sum up the last check to have:
```
def empty?
      return false if tags.any?
      return false if custom.any?
      user.empty?
end
```

File: lib/elastic_apm/context_builder.rb [hunk 4]
```diff
@@ -42,6 +42,16 @@ def apply_to_request(context, rack_env)
...
     # rubocop:enable Metrics/MethodLength, Metrics/AbcSize
 
+    def should_capture_body?(for_type)
+      option = config.capture_body
+
+      return true if option == 'all'
+      return true if option == 'transactions' && for_type == :transaction
+      return true if option == 'errors' && for_type == :error
+
+      false
```
Comment (simitt) [inline lib/elastic_apm/context_builder.rb]: The configuration doesn't change between requests, so you could store it as a class variable and then create another class var holding a hash with mappings for the type. Something along the lines of:
```
m = { transaction: true, error: true }
case config.capture_body
when 'off'
  m[:transaction] = false
  m[:error] = false
when 'transaction':
  m[:error] = false
when 'error':
  m[:transaction] = false
```

And then access it within `should_capture_body?(for_type)` with `m[for_type]`

File: lib/elastic_apm/error_builder.rb [hunk 5]
```diff
@@ -59,10 +50,20 @@ def add_stacktrace(error, kind, backtrace)
...
+
       error.transaction_id = transaction.id
       error.transaction = { sampled: transaction.sampled? }
+      error.trace_id = transaction.trace_id
+      error.parent_id = ElasticAPM.current_span&.id || transaction.id
+
+      return unless transaction.context
+
+      Util.reverse_merge!(error.context.tags, transaction.context.tags)
+      Util.reverse_merge!(error.context.custom, transaction.context.custom)
```
Comment (simitt) [inline lib/elastic_apm/error_builder.rb]: Why are you merging `tags` and `custom`, is this aligned with how other agents behave? I'd expect to find only the configured tags and customs if specifically set for an error.

File: lib/elastic_apm.rb (hunk 1 above)
Comment (mikker) [reply lib/elastic_apm.rb]: Backwards compatibility. Used to be 1 non-keyword arg. To keep it working with a deprecation warning, I have both.

File: lib/elastic_apm/context.rb (hunk 3 above)
Comment (mikker) [reply lib/elastic_apm/context.rb]: Right, I thought about that but liked the explicitness of this approach :)

File: lib/elastic_apm/context_builder.rb (hunk 4 above)
Comment (mikker) [reply lib/elastic_apm/context_builder.rb]: Oh, so memoize the method basically?

File: lib/elastic_apm/error_builder.rb (hunk 5 above)
Comment (mikker) [reply lib/elastic_apm/error_builder.rb]: Some do but it differs so I started https://github.com/elastic/apm/issues/62

File: lib/elastic_apm/context.rb (hunk 3 above)
Comment (simitt) [reply lib/elastic_apm/context.rb]: ok, that's fine with me.

File: lib/elastic_apm/context_builder.rb (hunk 4 above)
Comment (mikker) [reply lib/elastic_apm/context_builder.rb]: Skipping because the lookups aren't expensive in the first place 😊

**Comment by mikker** [comment]
Hmm, I think I'm missing `context.request` on errors using this branch. Will add a test and investigate.

**Pull request merged**